Feature/task 3

vaclavm opened the pull request.

vaclavm committed:
KA-276 Use MongoDB as repository + update Unity version

vaclavm committed:
KA-276 Introduce FormationService and IdentifierService for ToDo creation

Identifier is now created within the application (not by the database), and it is created inside new service.
Update tests to work with the new service.

vaclavm committed:
KA-276 Update ToDo model with dates and MongoDB identifier

vaclavm committed:
KA-276 Introduce DateTimeService

Inserting date into ToDo model during creation. Testing the correct creation.

vaclavm committed:
KA-276 Update ToDo in controller through IFormationService

vaclavm committed:
KA-276 Add correct connection string for MongoRepository

vaclavm committed:
KA-276 Update DependencyInjection test

Now working with reflection on ToDoList.Contracts and testing all Interfaces

vaclavm committed:
KA-276 Introduce IRetrieveService

IRetrieveService service retrieves items from repository, in my case ToDos. The service is used to get the ToDo, save one ToDo in cache and for checking if the ToDo exists.

vaclavm committed:
KA-276 Update DependencyInjection - not checking generics

vaclavm committed:
KA-276 Check IFormationService to IModificationService and add deletion

vaclavm committed:
KA-276 Introduce ViewModel and its validation in controller

ViewModel is add to prevent strict rules for client when PUT/POST. ViewModel converts to repository Model, as all other services works with the repository model.
Updated test to mock the convert method in ViewModel.

vaclavm committed:
KA-276 Remove the connectionStrings from repository

vaclavm committed:
KA-276 Add connectionString to .gitignore

vaclavm committed:
KA-276 Add model validation tests

vaclavm committed:
KA-276 Move services for ToDo to folder

vaclavm committed:
KA-276 Rename viewmodel Value to Text

vaclavm committed:
KA-276 Change creation to correctly return new Id

vaclavm committed:
KA-276 Updade .NET framework in ToDoList.Api

Tuple with names can be used

vaclavm committed:
KA-276 Add conditional exception for WebApi types

vaclavm committed:
KA-276 Cosmetic changes in code (lines, const, ...)

vaclavm committed:
KA-276 Change cache clearing and item deletion to be done by item, not only id

vaclavm committed:
KA-276 Add received calls tests

vaclavm committed:
KA-276 Add new folder in tests - to be same as project

vaclavm committed:
KA-276 Add tests from RetriveToDoService

vaclavm committed:
KA-276 Rename Retrive to Retrieve

vaclavm committed:
KA-276 Merge branch 'feature/task-2' into feature/task-3

vaclavm committed:
KA-276 Update project after merge with feature/task-2

vaclavm committed:
KA-276 Merge branch 'feature/task-2' into feature/task-3

vaclavm committed:
KA-276 Make IdentifierService internal

PetrSvirak commented on ToDoList/src/ToDoList.Api/ViewModels/ToDoViewModel.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        public string Text { get; set; }
+
+        public virtual ToDo Convert()
```
Only make methods virtual when you need them to be (or when creating a nuget library or something similarly ambitious).

Anyway, even if you needed the methods' to be virtual, never make `public` methods `virtual` - public methods represent (public) interface (~ way to communicate with) your class and should be as sparse as possible (similarly to the number of public classes - the less you have the less you have to maintain and cheaper the changes are (with less breaking changes). 
Virtual methods are opening your class to change allowing its successors to amend what the class does. If you needed the methods to be customizable, you would have to introduce protected virtual members for this task and call them from the public methods - this would allow you, for example, to implement some safety measures like wrap each protected member in a try-catch or whatever it is necessary to tame unknown code of the class' successors.

PetrSvirak commented on ToDoList/src/ToDoList.Api/ViewModels/ToDoViewModel.cs:
```diff
@@ -0,0 +1,29 @@
+            => new ToDo { Text = Text };
+
+        public virtual ToDo Convert(ToDo original)
+            => new ToDo { Id = original.Id, Text = Text, Created = original.Created, LastModified = original.LastModified };
```
Place each property assignment on a new line, please.

PetrSvirak commented on ToDoList/src/ToDoList.Api/ViewModels/ToDoViewModel.cs:
```diff
@@ -0,0 +1,29 @@
+
+            if (string.IsNullOrWhiteSpace(Text))
+            {
+                results.Add(new ValidationResult("The ToDo value must be set", new[] { "Text" }));
```
Use `yield return` and get rid of the `List` 😉

PetrSvirak replied:
Use `nameof` for field names

PetrSvirak replied:
Where are tests for this method?

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/Models/IConvertibleObject.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        T Convert();
+
+        T Convert(T original);
```
This is `Merge[With]` isn't it?

PetrSvirak replied:
I don't like this 2nd method - it moves business logic away from the business layer. How can model possibly know what parts of the `original` object are supposed to be reused?

For example `Put` will reuse everything but `LastModified`, yet you copy it anyway. Leave it up to the `Service` layer to decide what to do with the converted object it was provided with…

vaclavm replied:
Get your point, moved the resolution to modification service

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/Models/ToDo.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public class ToDo
+    {
+        [BsonId]
```
Data transfer object mustn't know a thing about `MongoDb` - otherwise, you are depending on mongo (and switch to, for example, MS SQL or DocumentDb would potentially break your code)

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/Models/ToDo.cs:
```diff
@@ -0,0 +1,20 @@
+        public DateTime LastModified { get; set; }
+
+        public override string ToString()
+            => $"Id: {Id}; Text: {Text}";
```
Your `ToString` misses something here 😉

PetrSvirak replied:
You should review your tests, if you didn't need `ToString` updated, you use them wrong 😛 (see below ;-))

PetrSvirak commented on ToDoList/src/ToDoList.Api.DependencyInjection/DependencyBootstrapper.cs:
```diff
@@ -0,0 +1,44 @@
+            => new DependencyBootstrapper(ContainerFactory.GetContainer())
+                .Register<ToDoList.Repository.DependencyInjection.DependencyRegister>()
+                .Register<ToDoList.Services.DependencyInjection.DependencyRegister>()
+                .Register<Api.Services.DependencyInjection.DependencyRegister>()
```
I have nothing against using complete namespaces here. However, either do it for all `DependencyRegister`s or for none, please.

vaclavm replied:
Using complete for all was my intention :)

PetrSvirak commented on ToDoList/src/ToDoList.Api.DependencyInjection/DependencyBootstrapper.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    public class DependencyBootstrapper
+    {
+        private static IContainer _container;
```
`Readonly`? Yes. `Static`? No, since you now change all potential resolvers you created by `CreateWebApiResolver` method - they should either be independent or just a single resolver should always be returned…

vaclavm replied:
Must have remained there from DI update in task 2.

PetrSvirak commented on ToDoList/test/ToDoList.Api.DependencyInjection.Tests/DependencyBootstrapperTests.cs:
```diff
@@ -0,0 +1,34 @@
+            var unregistredInterfaces = new [] { typeof(IContainer).FullName, typeof(IDependencyRegister).FullName };
+
+            // Act
+            DependencyResolver resolver = (DependencyResolver) DependencyBootstrapper.CreateWebApiResolver(routeHelper);
```
Do not change type in tests, especially not from the abstract (interface) to concrete (implementation) - it only means you wrote some code that is not testable. Rewrite the code in a way it will be testable - that is the solution 😉

PetrSvirak commented on ToDoList/test/ToDoList.Api.DependencyInjection.Tests/DependencyBootstrapperTests.cs:
```diff
@@ -0,0 +1,34 @@
+            // Assert
+            foreach (var interfaceType in interfaces.Where(type => !unregistredInterfaces.Contains(type.FullName)))
+            {
+                Assert.That(resolver.IsRegistered(interfaceType), Is.True, $"{interfaceType} don't have registred implementation");
```
Your assertion fails with first unregistered type, it would be great if all missing types were listed may the test fail…

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/DependencyInjection/IContainer.cs:
```diff
@@ -0,0 +1,22 @@
+
+        IContainer CreateChildContainer();
+
+        bool IsRegistered(Type serviceType);
```
Your application does not need this method, thus your contract should require it neither (do not add methods to your interfaces only for tests' sake).

vaclavm replied:
Yeah, I thought the same when I was simplifying the tests for DI as we talked not to resolve the services (GetService) in the tests.
But without it, I am not able to test registration of interfaces. I get use GetService method on resolver (I have already had that) but I would need to substitute HttpContext for LocationService and MongoDb for Repository service. 
So I am a little bit confused ...

PetrSvirak commented on ToDoList/src/ToDoList.DependencyInjection/Container/Container.cs:
```diff
@@ -0,0 +1,85 @@
+
+namespace ToDoList.DependencyInjection.Container
+{
+    internal class Container : IContainer
```
You really should look at the `DependencyResolver`'s implementation of `IDisposable` 😛

PetrSvirak commented on ToDoList/src/ToDoList.DependencyInjection/Container/Container.cs:
```diff
@@ -0,0 +1,85 @@
+                return _unityContainer.Resolve(serviceType);
+            }
+            catch (ResolutionFailedException exception) 
+                when (exception.Message.Contains(_webHttpSource) || exception.Message.Contains(_netHttpSource))
```
You should not rely on the `Message` containing the full type name. Especially since there is the `TypeRequested` property available in the `ResolutionFailedException` 😉

PetrSvirak replied:
Also, `contains` means anywhere in the text and being the type `ToDoList.System.Web.Http`, you would filter the exception while you shouldn't - use something like `exception.TypeRequested.StartsWith(_webHttpSource, StringComparison.OrdinalIgnoreCase)`, please.

PetrSvirak replied:
The code handling the condition repeats itself, try to sort this out, please.

vaclavm replied:
exception.TypeRequested only determines actual type (without namespaces), therefore I had to add ExcludedTypes property into IContainer to determine which exact namespaces could be excluded.

PetrSvirak commented on ToDoList/src/ToDoList.DependencyInjection/Container/Container.cs:
```diff
@@ -0,0 +1,85 @@
+            catch (ResolutionFailedException exception)
+                when (exception.Message.Contains(_webHttpSource) || exception.Message.Contains(_netHttpSource))
+            {
+                return new List<object>();
```
The `Enumerable.Empty<object>()` is what you should go with here.

PetrSvirak commented on ToDoList/src/ToDoList.DependencyInjection/Container/Container.cs:
```diff
@@ -0,0 +1,85 @@
+        public void RegisterType<TFrom, TTo>(LifetimeManager managerType)
+            where TTo: TFrom
+        {
+            switch (managerType)
```
Consider refactoring this so the method does not have 2 responsibilities (for picking up `LifetimeManager` and registering type using `RegisterType`.

PetrSvirak commented on ToDoList/src/ToDoList.Repository/MongoRepository.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace ToDoList.Repository
+{
+    internal class MongoRepository : IToDoRepository
```
This class is really nice as a whole 😃

PetrSvirak commented on ToDoList/src/ToDoList.Repository/MongoRepository.cs:
```diff
@@ -0,0 +1,42 @@
+    internal class MongoRepository : IToDoRepository
+    {
+        private readonly IMongoCollection<ToDo> _toToList;
+        private const string _collectionName = "ToDoItems";
```
R# complains about naming rule here - either fix the file or fix the rule (also, static members should precede instance ones) 😉

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/Services/IRetrieveService.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace ToDoList.Contracts.Services
+{
+    public interface IRetrieveService<T>
```
I believe it should be `Retrieval`

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,95 @@
+
+            var toDo = await _retrieveService.RetrieveOneAsync(id);
+            await _modificationService.DeleteAsync(toDo);
+            _retrieveService.ClearCache(toDo);
```
Nice thought with the cache 👍

PetrSvirak replied:
I can imagine these 3 lines being part of a facade. But were they remaining here, there is no reason not to use `Repository` directly to remove the item…

vaclavm replied:
I do not use Repository in the controller at all. To pass repository just for one method seems a little bit useless. I do not see any benefit in it.
Therefore I tried to make it easily readable: GetItem from retrieve service, Delete item in another service, Clear cache in retrieve service. It is not that much functionality for the action in controler.

vaclavm replied:
Because of your comment at RetrieveToDoService's method RetrieveAllAsync, I will use IToDoRepository.

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,95 @@
+                return NotFound();
+            }
+
+            return Ok(await _retrieveService.RetrieveOneAsync(id));
```
Please, move the method call to a standalone line, something like `var existingTodo = await _retrieveService.RetrieveOneAsync(id)`.

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,95 @@
+                return BadRequest(ModelState);
+            }
+
+            if (!await _retrieveService.IsInDatabaseAsync(id))
```
This is a bit problematic from the `REST` perspective since you do not use `id` in the newly created `todo`. Consider creating new items only for `id == Guid.Empty` and returning `NotFound` when (any other) identifier does not exist.

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,95 @@
+                return Created(location, newItem);
+            }
+
+            await _modificationService.UpdateAsync(toDoItem.Convert(await _retrieveService.RetrieveOneAsync(id)));
```
Your controller mustn't do this at all; simply pass the view model (as `IConvertibleObject<ToDo>`) and existing `ToDo` (or at least its `id`) to the service and let it do its magic, merging the two together and persisting them.

PetrSvirak replied:
Do not chain method calls, please. Each call should be on a standalone line. It can be tolerated in temporary code (such as `Task.FromResult(…)`, but should not be part of `production` one…

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/Services/IModificationService.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace ToDoList.Contracts.Services
+{
+    public interface IModificationService<T>
```
Do not presume you will ever create the same service again. If you ever did, you can use an interface like this, however, generics have its native place in the prevention of code repetition - since interfaces have no code, there is not much generics will bring you (unless you build some class that works with your generic interfaces).  It is highly unlikely that you will ever write code that reuses modification services (or any other services, true to be told). Please, get rid of the generics here and keep services' contracts simple.

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/Services/IModificationService.cs:
```diff
@@ -0,0 +1,14 @@
+
+        Task<T> UpdateAsync(T toUpdate);
+
+        Task<T> DeleteAsync(T id);
```
You really don't need this method and if you make it so you do, it really should accept just the `id` (GUID), not complete item…

PetrSvirak commented on ToDoList/src/ToDoList.Services/ToDoServices/ModificationToDoService.cs:
```diff
@@ -0,0 +1,54 @@
+
+        public async Task<ToDo> UpdateAsync(ToDo toUpdate)
+        {
+            toUpdate.LastModified = _dateTimeService.GetCurrentDateTime();
```
You should not modify the instance you got as a parameter. Especially since you return the same object type from the method.

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,95 @@
+            return StatusCode(HttpStatusCode.NoContent);
+        }
+
+        private async Task<(ToDo, string)> CreateToDoAsync(ToDoViewModel toDoItem)
```
Please, name the individual values in the tuple.

PetrSvirak commented on ToDoList/src/ToDoList.Api.Services/DependencyInjection/DependencyRegister.cs:
```diff
@@ -0,0 +1,25 @@
+            container.RegisterType(InjectHttpRequest);
+            container.RegisterType<IUrlLocationService, UrlLocationService>(LifetimeManager.Hierarchical);
+
+            var configuration = new ConnectionConfiguration { ConnectionString = ConfigurationManager.ConnectionStrings["MongoConnectionString"].ConnectionString };
```
At least make it a factory. But preferably, create an interface for this class and just implement it in this very assembly - it can read the connection string in its constructor (or eventually lazily on first getter access). It is definitely not the responsibility of the `Register` method (or any `IDependencyRegister` implementation) to know how exactly is a connection string obtained.

PetrSvirak commented on ToDoList/test/ToDoList.Api.Services.Tests/UrlLocationServiceTests.cs:
```diff
@@ -0,0 +1,48 @@
+            httpRequestMessage.Properties[System.Web.Http.Hosting.HttpPropertyKeys.HttpConfigurationKey] = configuration;
+            
+            _webApiRoutes = Substitute.For<IWebApiRoutes>();
+            _locationService = new UrlLocationService(new UrlHelper(httpRequestMessage), _webApiRoutes);
```
Just mock `UrlHelper`, do not mock all route configuration. `Substitute.For<UrlHelper>` and set what its `Route` function should return for given parameters - no need to meddle with routes and `HttpRequestMessage` at all…

PetrSvirak commented on ToDoList/test/ToDoList.Api.Services.Tests/UrlLocationServiceTests.cs:
```diff
@@ -0,0 +1,48 @@
+        public void GetNewResourceLocation_ReturnsCorrectLocation()
+        {
+            // Arrange
+            Guid guid = Guid.NewGuid();
```
This makes the test unrepeatable. Use some predefined GUIDs (everywhere), please.

PetrSvirak commented on ToDoList/src/ToDoList.Services/ToDoServices/RetrieveToDoService.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        public async Task<IEnumerable<ToDo>> RetrieveAllAsync()
+            => await _toDoRepository.GetToDosAsync();
```
This method really is useless and it is connected to the other methods you have here in no way.

I told you that you can create facades if you think it is helpful - this still stands - but, do not mix facades with services themselves - facades must be isolated in terms of combining services together and (not) introduce service-level logic within them. 

You can create a facade for item retrieval operations that would basically combine retrieval service with some repository methods, however, you cannot merge all into a single class with retrieval service's logic (cache in this case).

vaclavm replied:
Ok

PetrSvirak commented on ToDoList/test/ToDoList.Services.Tests/ToDoServices/ModificationServiceTests.cs:
```diff
@@ -0,0 +1,107 @@
+            Assert.That(response.Text, Is.EqualTo(expectedToDo.Text), "ToDo item is not as expected");
+            Assert.That(response.Id, Is.EqualTo(guid), "Id is not as expected");
+            Assert.That(response.Created, Is.EqualTo(currentTime), "Created date is not as expected");
+            Assert.That(response.LastModified, Is.EqualTo(currentTime), "Last modified date is not as expected");
```
You have `ToDoComparer` for this task. Definitely, do reuse it for services' tests

PetrSvirak commented on ToDoList/test/ToDoList.Services.Tests/ToDoServices/ModificationServiceTests.cs:
```diff
@@ -0,0 +1,107 @@
+        public async Task CreateToDoAsync_ToDoIsSetCorrectly()
+        {
+            // Arrange
+            var expectedToDo = new ToDo { Text = "Test item" };
```
This is actually not an `expectedToDo`, this is todo you get as an input - it might actually be more appropriate to receive `IConvertibleObject<ToDo>` as an argument…

PetrSvirak commented on ToDoList/test/ToDoList.Services.Tests/ToDoServices/ModificationServiceTests.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        [Test]
+        public async Task UpdateToDoAsync_OriginalFieldsUnchanged()
```
What additional value does this test brings when compared with `UpdateToDoAsync_LastModifiedIsUpdated`?

vaclavm replied:
Nothing, deleted :)

PetrSvirak commented on ToDoList/test/ToDoList.Services.Tests/ToDoServices/RetrieveToDoServiceTests.cs:
```diff
@@ -0,0 +1,116 @@
+        }
+
+        [Test]
+        public async Task IsInDatabaseAsync_NotExists_ReturnsFalse()
```
Question is whether you shouldn't cache nonexistent items as well to prevent re-scanning database for an `id` that doesn't exist and you already know it…

vaclavm replied:
I don't think so as the item can have been created but the service would still return false.
It is possible to cache nonexisting id of item but would require another logic that I don't find relevant in our case.

PetrSvirak commented on ToDoList/src/ToDoList.Services/DateTimeService.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    internal class DateTimeService : IDateTimeService
+    {
+        public DateTime GetCurrentDateTime()
```
I miss tests for this implementation…

PetrSvirak commented on ToDoList/src/ToDoList.Services/IdentifierService.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    internal class IdentifierService : IIdentifierService
+    {
+        public Guid GenerateIdentifier()
```
I miss tests for this implementation…

PetrSvirak commented on ToDoList/test/ToDoList.Api.Tests/Controllers/ToDosControllerTests.cs:
```diff
@@ -0,0 +1,289 @@
+            const int itemIndex = 2;
+            var expectedToDo = _toDoList.ElementAt(itemIndex);
+            string location = $"todos/{expectedToDo.Id}";
+            var postToDo = Substitute.For<ToDoViewModel>();
```
new ToDoViewModel { Text = <what-ever> } seems to be more fitting that substitution here - after all, was the `Convert` more complex, you would test it and could use it anyway. 

Note: Problem with your substitute is that it actually lies a bit since it returns complete `ToDo` item, not only item with the `Text` property filled.

PetrSvirak reviewed (changes requested):
OK, it is not bad, but it needs some attention and a general look at the solution as a whole.

I would like you to think about bootstrap (`DependencyRegister`) as an entry point of your assemblies - the only class that has public modifier and only class that has the right to be in the root folder of an assembly (`DependencyInjection` has `ContainerFactory`, `Api.DependencyInjection` has `DependencyBootstrapper`, and `Api` has `global.asax` - but otherwise, it should always be an implementation of the `IDependencyRegister` that welcomes new programmer in an unknown DLL of yours). 

All other implementations should be under a folder whatever assembly they are in. Since you have `Contract`, you can use folder structure present there to aid you with folder names. It will give you a bit funny namespaces from time to time (e.g. `ToDoList.Services.Services` or `ToDoList.Repository.Repositories`), but that is quite OK since these namespaces will never leave the assembly with the implementation of a contract and it will be easy to understand which part of contracts is implemented in which assembly.

Services (and eventually facades you might decide to create (or not)) should not be in a single stack in the `Contracts` assembly (just like they are now) but grouped in the folders describing relating factor of their nature. For example, `Todo`-related services could have their folder and there are these special wrappers that isolated static method calls and could also be united under one folder (e.g. `StaticWrappers` or something).

Do not limit yourself by naming everything with `Service` suffix - for example, `IWebApiRoutes` is much better name than `IRoutesService` (but file somehow avoided renaming). The `IIdentifierService` might (or might not) be an `IdentifierGenerator`. The `IDateTimeService` could (but mustn't) be an `ITimeProvider` or even an `ITimeMachine`. The `IUrlLocationService` always wanted (or not) to be `IUrlLocator` but never dared to admit it. I mean, `Service` suffix is always a fallback, but if you can come up with a better name, just give it a though and a go, observe how it works in the bigger picture and eventually adopt the new name.

Once at a time, it is good to go through all the solution and ask yourself whether names and the way the code is assembled together make sense at the first glance - whether renaming a method here and a local variable there or introducing few extensions methods over here wouldn't "tell the story" of your API in a clearer way.

One small notice: when naming tests, please, stick with `<MethodName>_<ParametersDescription>_<OutcomeDescription>` - you seem to sometimes omit parameter description or provide just a vague description on expected test/tested method outcome.

While some things I just wrote above are not exactly measurable, notes below should be and need your attention as well the complete picture your code paints:

vaclavm committed:
KA-276 Remove Convert(T) method from viewmodel

What should be copy from one to another item is now in modification service.
Also tests updated.

vaclavm committed:
KA-276 Add new tests from services

vaclavm committed:
KA-276 Add tests for Validate and Convert in ToDoViewModel

vaclavm committed:
KA-276 Update ToString method

vaclavm committed:
KA-276 Change dispose and exception handling to type

vaclavm committed:
KA-276 Make DependencyBootstrapper instantiable and create resolver in its method

As DependencyBootstrapper us now instantiable, it can be easily tested. To test it properly I have created DummyContainer out of IContainer, that is used in DependencyBootstrapper.
Moved ExcludedTypes to API.DI, as I did not want to use .NET/.WEB dlls in base DI project.

vaclavm committed:
KA-276 Make cosmetic changes in controler

vaclavm committed:
KA-276 Create interface for connection configuration

vaclavm committed:
KA-276 Delete LifetimeManager and make registration for singleton separate

vaclavm committed:
KA-276 Create Test.Utils project + use Guid.Parse in tests

Move ToDoComparer and ActionExtension to the project so it can be used in other Tests.
Using Guid.Parse with specific guid to prevent problems with Guid.NewGuid.

vaclavm committed:
KA-276 Simplify UrlLocationServiceTests

vaclavm committed:
KA-276 Fix R# warning of naming

vaclavm committed:
KA-276 Delete InternalVisibleTo

vaclavm committed:
KA-276 Transfer folder structure in ToDoList.Repository

vaclavm committed:
KA-276 Transfer folder structure in ToDoList.Services

vaclavm committed:
KA-276 Transfer folder structure in ToDoList.Api.Services

vaclavm committed:
KA-276 Transfer folder structure in ToDoList.Api.Services.Tests

vaclavm committed:
KA-276 Add tests from ConnectionConfigurationService

vaclavm committed:
KA-276 Update travis with new test .dll

vaclavm committed:
KA-276 Rename interfaces to proper name

vaclavm committed:
KA-276 Rename IRetrieveService to IRetrievalService in code

vaclavm committed:
KA-276 Rename RetrieveToDoService to RetrievalToDoService

vaclavm committed:
KA-276 remove xbuild from .travis

vaclavm committed:
KA-276 return xbuild

vaclavm committed:
KA-276 Transfer folder structure in ToDoList.Contracts

 Rename files with services to new names
 Adjust namespaces for new providers
 Refactor solution to ensure new naming

vaclavm committed:
KA-276 Transform structure in ToDoList.Api to correspond with the Contracts

Rename and move RoutesHelper to WebApiRoutes

vaclavm committed:
KA-276 Transfer structure in ToDoList.Services to correspond with the Contracts

Move files from services to providers, move files from todoservices to services
Refactor solution for provide correct names

vaclavm committed:
KA-276 Transform structure in ToDoList.Repository to correspond with the Contracts

vaclavm committed:
KA-276 Move IConnectionConfiguration to proper folder

vaclavm committed:
KA-276 Transform structure in ToDoList.Api.Services to correspond with the Contracts

Rename and move the files to proper folders
Refactor the solution using proper naming

vaclavm committed:
KA-276 Transform ToDoList.Api.Services.Tests to correspond with the origin project

vaclavm committed:
KA-276 Transform ToDoList.Services.Tests to correspond with the origin project

vaclavm committed:
KA-276 Refactor generic interfaces to non-generic

vaclavm committed:
KA-276 Rename generic interfaces to proper names

vaclavm committed:
KA-276 Make public constructor or DI

vaclavm committed:
KA-276 Remove MongoRepository reference from contracts

vaclavm committed:
KA-276 Check all interfaces in test

PetrSvirak commented on ToDoList/test/ToDoList.Api.DependencyInjection.Tests/DependencyBootstrapperTests.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            // Arrange
+            var routeHelper = new WebApiRoutes();
+            var assembly = Assembly.Load("ToDoList.Contracts");
```
Rather use something like `typeof(IDependencyRegister).Assembly` – it is strongly typed and enforces reference to the (contracts) assembly in tests. 😛

PetrSvirak commented on ToDoList/test/ToDoList.Api.DependencyInjection.Tests/DependencyBootstrapperTests.cs:
```diff
@@ -0,0 +1,44 @@
+                {
+                    ungeristredInterfaces.Add(interfaceType.ToString());
+                }
+            }
```
What about LINQ:
````
var ungeristredInterfaceNames = interfaces
    .Where(type => !unregistredInterfaces.Contains(type.FullName))
    .Where(@interface => resolver.GetService(@interface) == null)
    .Select(@interface => @interface.FullName)
    .ToArray();
````
😛

vaclavm replied:
Nice!

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,106 @@
+                return BadRequest(ModelState);
+            }
+
+            (ToDo newItem, string location) = await CreateToDoAsync(toDoItem);
```
Consider using `var`s → `(var newItem, var location) = ` or `var` → `var creation = … creation.NewToDo, creation.Location`…

PetrSvirak commented on ToDoList/src/ToDoList.Api/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,106 @@
+            return StatusCode(HttpStatusCode.NoContent);
+        }
+
+        private async Task<(ToDo newToDo, string location)> CreateToDoAsync(ToDoViewModel toDoItem)
```
The `newTodo` and `location` are properties and this is C# 😉

PetrSvirak commented on ToDoList/src/ToDoList.Api/ViewModels/ToDoViewModel.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+        {
+            if (string.IsNullOrWhiteSpace(Text))
```
Usually, if a type is used like a class with members, a capital first letter is preferred (`string name = "…";`, but `String.IsNullOrWhiteSpace`)…

PetrSvirak commented on ToDoList/src/ToDoList.Api.DependencyInjection/DependencyBootstrapper.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public IDependencyResolver CreateWebApiResolver(IWebApiRoutes webApiRoutes)
+        {
+            return Register<ToDoList.Repository.DependencyRegister>()
```
Missing lambda here 😛

PetrSvirak commented on ToDoList/src/ToDoList.Api.Services/Providers/ConnectionConfiguration.cs:
```diff
@@ -0,0 +1,13 @@
+        private string _connectionString;
+
+        public string ConnectionString 
+            => _connectionString ?? (_connectionString = ConfigurationManager.ConnectionStrings["MongoConnectionString"].ConnectionString);
```
Haven't you been working on an EMS product? 😛 

It is an equivalent of 
```
private readonly Lazy<string> _connectionString = new Lazy<string>(() => ConfigurationManager.ConnectionStrings["MongoConnectionString"].ConnectionString);

public string ConnectionString => _connectionString.Value;
```

However, the `Lazy` implementation is thread-safe as a bonus.

BTW: The following would also do:
```
public string ConnectionString { get; }

public ConnectionConfiguration()
{
    ConnectionString = ConfigurationManager.ConnectionStrings["MongoConnectionString"].ConnectionString;
}
```

vaclavm replied:
Thanks for the Lazy hint ;-)

PetrSvirak commented on ToDoList/src/ToDoList.Api.Services/Providers/Locator.cs:
```diff
@@ -0,0 +1,21 @@
+            _webApiRoutes = webApiRoutes;
+        }
+
+        public string GetNewResourceLocation(Guid id)
```
I like the new overall naming. However, this isn't just any resource - it is `ToDo` if you got a new resource that would require, I don't know, a `Guid` id and a `string` username, you would surely create a new method in the `Locator`…

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/DependencyInjection/IContainer.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    public interface IContainer : IDisposable
+    {
+        IEnumerable<string> ExcludedTypes { get; set; }
```
This gives quite a control to the classes using the container. It is a faster solution to write than creating your own exception (pushing the problem to the `IDependencyResolver` where it technically belongs) or making the excluded types a dependency passed through the constructor (making it a parameter of the `ContainerFactory` and passing an array of the strings to it in the `DependencyBootstrapper`), but it just does not feel right – container looks too unstable when each `IDependencyRegister` can reset its `ExcludedTypes`…

PetrSvirak replied:
You should actually store a `Set<T>` internally since you want to search through it quite often and each type needs to be present but once…

PetrSvirak replied:
Never leave your `Enumerable<T>`s `null` – collections in C# are designed to follow [Null-Object pattern](https://en.wikipedia.org/wiki/Null_object_pattern). The least you could do is append ` = Enumerable.Empty<string>();` here…

PetrSvirak commented on ToDoList/src/ToDoList.Api.DependencyInjection/DependencyBootstrapper.cs:
```diff
@@ -0,0 +1,68 @@
+                nameof(IContentNegotiator),
+                nameof(IExceptionHandler),
+                nameof(IModelValidatorCache),
+                nameof(ModelMetadataProvider),
```
Preferably use `typeof(T).FullName` so anybody can define a `IHostBufferPolicySelector` or any of the others in their assembly…

PetrSvirak commented on ToDoList/src/ToDoList.Services/Services/ModificationToDoService.cs:
```diff
@@ -0,0 +1,47 @@
+
+            toUpdate.Text = newToDo.Text;
+            toUpdate.LastModified = _timeProvider.GetCurrentDateTime();
+            await _toDoRepository.ChangeToDoAsync(toUpdate);
```
To be consistent with [`CreateAsync`](https://github.com/vaclavm/kentico-onboarding-cs/pull/7/files/0cff8c1e36248f4a3daca4094565a82274182b8b..HEAD#diff-363caedfd63743d5fa9fcdd587cb7e6dR23), the empty line should be above, not below the `_toDoRepository` 😛

PetrSvirak commented on ToDoList/test/ToDoList.Api.DependencyInjection.Tests/Container/DummyContainer.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    internal class DummyContainer : IContainer
+    {
+        private List<Tuple<string, string>> _registrations = new List<Tuple<string, string>>();
```
Naming the individual tuples here would help both the reader and your code - you would avoid `.Item1` and `.Item2` here and there 😛

PetrSvirak commented on ToDoList/test/ToDoList.Api.Services.Tests/Providers/ConnectionConfigurationTests.cs:
```diff
@@ -0,0 +1,19 @@
+    public class ConnectionConfigurationServiceTests
+    {
+        [Test]
+        public void ConnectionString()
```
It would be nice if the test was self-contained, but that is just a detail (that would either require mocking `ConfigurationManager` or hiding it behind an interface similar to the `DateTime` or `Guid`) 😛 I appreciate you wrote such kind of test at all though 😃

vaclavm replied:
Thanks for appreciations! I will leave the test how it is, if you don't mind :)

PetrSvirak reviewed (changes requested):
Below are mostly details and could have been fixed in the next task (or none). The only thing I actually selected "Request changes" are the `ExcludedTypes` – or more precisely the way they propagate and integrate (the `IsWebOrNetException` method just does not seem to be named right and/or in the right place). They seem to be added in a bit of hurry from the design perspective (at least to me) and are the only significant (some might say meaningful 😆) flaw in your solution right now. Very good job 👍 

Note: Especially the `ConnectionConfigurationTests.cs` is something I do NOT expect you to solve 😉

vaclavm committed:
KA-276 Move exception handling to DependencyResolver

Create custom exception that is not dependent on unity exception, therefore it can be used in other parts of system.

vaclavm committed:
KA-276 Simplify tests and use named tuples

vaclavm committed:
KA-276 Refactor code - cosmetic changes + rename of GetNewResourceLocation

vaclavm committed:
KA-276 Change tuple properties naming
